buffer/src/main/java/io/netty/buffer/AbstractReferenceCountedByteBuf.java

```diff
@@ -59,18 +59,12 @@ public ByteBuf retain(int increment) {
         return retain0(checkPositive(increment, "increment"));
     }
 
-    private ByteBuf retain0(int increment) {
-        for (;;) {
-            int refCnt = this.refCnt;
-            final int nextCnt = refCnt + increment;
-
-            // Ensure we not resurrect (which means the refCnt was 0) and also that we encountered an overflow.
-            if (nextCnt <= increment) {
-                throw new IllegalReferenceCountException(refCnt, increment);
-            }
-            if (refCntUpdater.compareAndSet(this, refCnt, nextCnt)) {
-                break;
-            }
+    private ByteBuf retain0(final int increment) {
+        int oldRef = refCntUpdater.getAndAdd(this, increment);
+        if (oldRef <= 0 || oldRef + increment < oldRef) {
+            // Ensure we don't resurrect (which means the refCnt was 0) and also that we encountered an overflow.
+            refCntUpdater.getAndAdd(this, -increment);
```

Comment:
Why not ```refCntUpdater.getAndSet(this, oldRef);```?

Reply:
Because it could have changed due to another thread updating it.

Reply:
@carl-mastrangelo Oh, of course. Agree. 👍